docker/build.bzl

```diff
@@ -11,7 +11,35 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Rule for building a Docker image."""
+"""Rule for building a Docker image.
+
+In addition to the base docker_build rule, we expose its constituents
+(attr, outputs, implementation) directly so that others may expose a
+more specialized build leveraging the same implementation.  The
+expectation in such cases is that users will write something like:
+
+  load(
+    "@io_bazel_rules_docker//docker:build.bzl",
+    _build_attrs="attrs",
```

Comment:
There is an internal doc (/cc @laurentlb for making that doc public) about cleaning up namespace of native symbols. A consequence would be that we would write java.merge instead of java_common.merge, with `java` being a symbol loaded from an extension. That's the reason for struct.

For named providers this is just something to consider since this should be the new way.